### PR TITLE
Cache the last string's hash we computed in our filepath map

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Drops CPU cost in VS during solution load from:

image

to

image

### CODE REVIEW

**Comment by CyrusNajmabadi** [comment]
@ToddGrun ptal.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CachingFilePathComparer.cs [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+                if (lockTaken)
+                    _lock.Exit();
+            }
+
+            if (ReferenceEquals(lastString, obj))
+                return lastHashCode;
+
+            // Specialized impl of OrdinalIgnoreCase.GetHashCode that is faster for the common case of an all-ASCII
+            // string. Falls back to normal OrdinalIgnoreCase.GetHashCode for the uncommon case.
+            var hashCode = GetNonRandomizedHashCodeOrdinalIgnoreCase(obj);
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CachingFilePathComparer.cs:75]: >GetNonRandomizedHashCodeOrdinalIgnoreCase [](http://example.com/codeflow?start=27&length=41)

ok, dumb question: why not just call into obj.GetHashcode?

Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CachingFilePathComparer.cs:75]: Case sensitivity, I suppose. There isn't something exposed that does this?

Comment (ToddGrun) [reply src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CachingFilePathComparer.cs:75]: And, I now see the comment right above it. Never mind.

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CachingFilePathComparer.cs [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+
+        public int GetHashCode([DisallowNull] string obj)
+        {
+            string? lastString;
+            int lastHashCode;
+
+            var lockTaken = false;
+            try
+            {
+                _lock.Enter(ref lockTaken);
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CachingFilePathComparer.cs:60]: >_lock.Enter(ref lockTaken); [](http://example.com/codeflow?start=16&length=27)

Should this check lockTaken before assigning to the locals?

File: src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CachingFilePathComparer.cs [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+                lastString = _lastString;
+                lastHashCode = _lastHashCode;
+            }
+            finally
+            {
+                if (lockTaken)
+                    _lock.Exit();
+            }
+
+            if (ReferenceEquals(lastString, obj))
```
Comment (ToddGrun) [inline src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.CachingFilePathComparer.cs:70]: >if (ReferenceEquals(lastString, obj)) [](http://example.com/codeflow?start=12&length=37)

Would it make things ugly to check _lastString against obj and only taking the lock if ReferenceEqual?

**Review by ToddGrun — APPROVED** [review]

**Pull request merged**